src/fd.cpp

```diff
@@ -110,19 +159,99 @@ static int wait_thread(void *arg) {
     std::fill(args->results->begin(), args->results->end(), NA_INTEGER);
   }
 
+  return 0;
+
+}
+
+static int wait_thread_single(void *arg) {
+
+  tct_thrd_detach(tct_thrd_current());
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  if (wait_on_fds(args))
+    return 1;
+
   callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
 
   return 0;
 
 }
 
+static int wait_thread_persistent(void *arg) {
+
+  tct_thrd_detach(tct_thrd_current());
+  thread_active->store(true); // atomic allows update prior to acquiring lock
+
+  if (cv.lock()) THREAD_RETURN(1);
+  if (cv.signal()) { cv.unlock(); THREAD_RETURN(1); } // signal to sync with main thread
+  cv.busy(false);
+  while (!cv.busy()) {
+    if (cv.wait()) { cv.unlock(); THREAD_RETURN(1); }
+  }
+  if (cv.unlock()) THREAD_RETURN(1);
+
+  while (1) {
+
+    if (!thread_active->load()) break; // set by cv destructor on exit
+
+    std::shared_ptr<ThreadArgs> args = *thread_args;
+
+    if (wait_on_fds(args)) {
+      thread_args.reset(); // cancelled so just reset
+    } else {
+      callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(thread_args.release()), 0, args->loop);
+    }
+
+    if (cv.lock()) THREAD_RETURN(1);
+    cv.busy(false);
+    while (!cv.busy()) {
+      if (cv.wait()) { cv.unlock(); THREAD_RETURN(1); }
+    }
+    if (cv.unlock()) THREAD_RETURN(1);
+
+  }
+
+  cv.destroy(); // must do this from here rather than cv destructor to avoid deadlock
+  return 0;
+
+}
+
 static int execLater_launch_thread(std::shared_ptr<ThreadArgs> args) {
 
   std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(new std::shared_ptr<ThreadArgs>(args));
 
+  if (!thread_active->load()) {
+
+    tct_thrd_t thr;
+    if (tct_thrd_create(&thr, &wait_thread_persistent, NULL) != tct_thrd_success)
+      return 1;
+
+    if (cv.lock()) return 1;
+    while (!thread_active->load()) { // thread can update this without acquiring lock
+      if (cv.wait()) { cv.unlock(); return 1; } // wait for signal from thread
+    }
+    if (cv.unlock()) return 1;
+
+  }
+
+  do {
+
+    if (cv.lock()) return 1;
+    if (cv.busy()) { cv.unlock(); break; } // busy so create new single thread
+
+    cv.busy(true);
```

Comment:
I wonder if it would simplify things to collapse `thread_active->load` and `cv.busy` into one non-atomic variable with these states: 1. thread not started, 2. thread idle, 3. thread busy, 4. thread (should) shutdown. And it's always accessed under a mutex, and always cv notify when it's written to. The persistent thread does `while (state == THREAD_IDLE) cv.wait()`.

I really, really dislike having some state protected by a mutex, and other (tightly related) state not being protected. Even if the logic is perfectly correct today, and even if it's easy for you to reason about perfectly, it will be very difficult to keep correct in the future.

Reply:
Thinking about this, I totally agree that the atomic flag, which I thought greatly simplified things, was actually making my condition variable logic non-standard. Doing everything under lock is likely going to be the way forward.